text/0000-fn-lifetime.md

```diff
@@ -0,0 +1,86 @@
+- Feature Name: fn lifetime
+- Start Date: 2017-01-02
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Add a `'fn` lifetime that is bound to the scope of the body of the current
+innermost function or closure.
+
+# Motivation
+[motivation]: #motivation
+
+Doing this will enable us to declare that some values live longer than their
+containing scope, allowing us to "allocate" recursive data structures (e.g.
+linked lists) on the stack, as in:
+
+```rust
+struct Node<T> {
+    data: T,
+    next: Option<&'fn Node>,
```

Comment:
@Ericson2314 If we had nested generics (which we don't), the former requires injecting an extra parameter, which is supplied by uses in the surrounding function, which is my second point.
There is no need to allow it in this form, and I don't think @whitequark was thinking of doing anything about type declarations (I sure wasn't), only *borrowing expressions*.

That is, `&'fn Node { ... }` to allocate one in the current stack frame, dynamically.
Oh btw you can't infer this allocation thing from the type, it has to *syntactically* be in the borrow expression for it to be usable to place it in the right scope accordingly.
Which leads to... drops. How would one even drop such a thing? A linked list of dynamic allocations and their associated drop glue? This might actually *require* something that goes with placement-in to allocate the raw space on the stack, and to be used with custom-allocator-enabled containers.